sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -240,6 +242,35 @@ public void setCoderRegistry(CoderRegistry coderRegistry) {
      * visited.
      */
     public void visitValue(PValue value, TransformTreeNode producer);
+
+    /**
+     * Control enum for indicating whether or not a traversal should process the contents of
+     * a composite transform or not.
+     */
+    public enum Recurse {
+      ENTER_TRANSFORM,
+      DO_NOT_ENTER_TRANSFORM;
+    }
+
+    /**
+     * Default no-op {@link PipelineVisitor} that enters all composite transforms.
+     * User implementations can override just those methods they are interested in.
+     */
+    public class Defaults implements PipelineVisitor {
+      @Override
+      public Recurse enterCompositeTransform(TransformTreeNode node) {
```

Comment:
I'm a bit curious about this API. I'm wondering if it would make more sense to have:

`beforeTransform(...)`
`visitTransform(...)`
`afterTransform(...)`

(I think that's approximately what the `enterCompositeTransform` and `leaveCompositeTransform` are doing).